doc/ir-semantics.text

```diff
@@ -2,34 +2,42 @@
 
 x        variables
 
-P ::= Segment  active   ↦ S, (Ds ↦ S)*   program:            active version and others
+P ::= Function main    ↦ F, (Df ↦ F)*    program:            main function and other functions
+F ::= Segment  active  ↦ S, (Ds ↦ S)*    function:           active version and others
 S ::= start ↦ i,          I              segment:            instruction stream with dedicated start
 I ::= (l ↦ i)*                           instruction stream: labeled instructions
 
 Ds ::= Segment  s                  segment declaration
 
+f  function name
 s  optimization version
 l  labels
 
+main      main function
 active    active segment
 start     start label
 
 a  addresses
 
 i ::=    instructions
-| const x = c                   constant variable
-| mut x = c                     mutable variable
-| x <- c                        assignment
+| const x = ce                  constant variable
+| mut x = ce                    mutable variable
+| x <- ce                       assignment
 | branch e l₁ l₂                conditional
 | goto l                        goto
 | print e                       print
+| return e                      return
```

Comment:
Ok I see this is causing some confusion.

CC is just an execution context, the only thing it really helps us is in applying eval to one instruction in the right order.

C is the continuation and it contains just the label, not the actual expression. but I guess you are right that by storing the expression in the continuation it will be a lot more intuitive. Let me try and do that.